lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+    } catch {
+        $driverCertificate = Get-JujuCharmConfig -Scope "ovs-certificate-url"
+        if ($driverCertificate) {
+            $CertificatePath = Get-OVSCertificate
+            Import-OVSCertificate $CertificatePath
+        }
+        else {
+            Write-JujuLog "No OVS driver certificate specified. Be sure to use an offically signed driver."
+        }
+        $InstallerPath = Get-OVSInstaller
+        Install-OVS $InstallerPath
+    }
+    if(!(Test-Path $OVS_VSCTL)){
+        Throw "Could not find ovs-vsctl.exe in location: $OVS_VSCTL"
+    }
+}
+
+
+function Enable-OVSExtension {
+    $ext = Get-OVSExtStatus
+    if (!$ext){
+       Throw "Cannot enable OVS extension. Not installed"
+    }
+    if (!$ext.Enabled) {
+        Enable-VMSwitchExtension $OVS_EXT_NAME $ext.SwitchName
+    }
+    return $true
+}
+
+
+function Get-OVSExtStatus {
+    $br = Get-JujuVMSwitchName
+    Write-JujuInfo "Switch name is $br"
+    $ext = Get-VMSwitchExtension -VMSwitchName $br -Name $OVS_EXT_NAME
+
+    if (!$ext){
+        Write-JujuInfo "Open vSwitch extension not installed"
+        return $null
+    }
+
+    return $ext
+}
+
+
+function Enable-Service {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$ServiceName
+    )
+
+    Get-Service $ServiceName | Set-Service -StartupType Automatic
+}
+
+
+function Set-HyperVMACS {
+    $b1 = "0x{0:x}" -f (get-random -minimum 1 -maximum 255)
+    $b2 = "0x{0:x}" -f (get-random -minimum 1 -maximum 255)
+    $reg = "HKLM:\Software\Microsoft\Windows NT\CurrentVersion\Virtualization"
+    set-ItemProperty -path $reg -name minimummacaddress -type binary -value ([byte[]](0x00,0x15,0x5d,$b1,$b2,0x00))
+    set-ItemProperty -path $reg -name maximummacaddress -type binary -value ([byte[]](0x00,0x15,0x5d,$b1,$b2,0xff))
+}
+
+
+function Enable-OVS {
+    Start-ExternalCommand { pip install -U "ovs" } -ErrorMessage "Failed to install $ovs_pip"
+    
+    Start-ExternalCommand { cmd.exe /c "sc triggerinfo ovs-vswitchd start/strcustom/6066F867-7CA1-4418-85FD-36E3F9C0600C/VmmsWmiEventProvider" } -ErrorMessage "Failed to modify ovs-vswitchd service."
+    Start-ExternalCommand { cmd.exe /c "sc config ovs-vswitchd start=demand" } -ErrorMessage "Failed to modify ovs-vswitchd service."
+
+    Enable-OVSExtension
+
+    Enable-Service "ovsdb-server"
+    #Enable-Service "ovs-vswitchd"
+
+    Start-Service "ovsdb-server"
+    #Start-Service "ovs-vswitchd"
+}
+
+
+function Ensure-InternalOVSInterfaces {
+    $ifIndex = Get-CharmState -Namespace "novahyperv" -Key "dataNetworkIfindex"
+    $lip = Get-CharmState -Namespace "novahyperv" -Key "local_ip"
+    $ifName = (Get-NetAdapter -ifindex $ifIndex).Name
+    $lip_mask = Get-CharmState -Namespace "novahyperv" -Key "local_ip_mask"
+
+    Invoke-JujuCommand -Command @($ovs_vsctl, "--may-exist", "add-br", "juju-br")
+    Invoke-JujuCommand -Command @($ovs_vsctl, "--may-exist", "add-port", "juju-br", $ifName)
+
+    Restart-Service "ovs-vswitchd"
```

Comment:
as far as i can tell sometimes juju-br adapter won't enable if ovs-vswitchd is not restarted